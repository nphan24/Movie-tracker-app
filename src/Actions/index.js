export const postMovies = movies => ({
  type: 'POST_MOVIES',
  movies
});

export const addUser = user => ({
  type: 'ADD_USER',
  user
});

export const signInAction = user => ({
  type: 'SIGN_IN',
  user
});

export const validSignIn = error => ({
  type: 'VALID_SIGNIN',
  error
});

export const invalidSignIn = error => ({
  type: 'INVALID_SIGNIN',
  error
});

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
});
