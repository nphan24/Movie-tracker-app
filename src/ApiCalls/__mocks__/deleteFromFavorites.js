export const deleteFromFavorites = jest.fn();
