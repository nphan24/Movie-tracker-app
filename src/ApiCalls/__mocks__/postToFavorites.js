export const postToFavorites = jest.fn();
