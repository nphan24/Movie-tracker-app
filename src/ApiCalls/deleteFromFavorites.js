export const deleteFromFavorites = async (movieid, userid) => {
  try {
    await fetch(`api/users/${userid}/favorites/${movieid}`, {
      method: 'DELETE'
    });
  } catch (error) {
    throw new Error(error.message);
  }
};
