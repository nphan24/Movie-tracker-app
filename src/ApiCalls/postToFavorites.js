const postToFavorites = async (movie) => {
  console.log(movie);
  const { movie_id, user_id, title, poster_path, release_date, vote_average, overview } = movie;
  try {
    const response = await fetch('api/users/favorites/new', {
      method: 'POST',
      body: JSON.stringify({movie_id, user_id, title, poster_path, release_date, vote_average, overview}),
      headers: {
        'Content-Type' : 'application/json'
      }
    });
    console.log(response);
    const favId = await response.json();
    return favId;
  } catch (errs) {
    throw new Error(errs.message);
  }
}

export { postToFavorites };



