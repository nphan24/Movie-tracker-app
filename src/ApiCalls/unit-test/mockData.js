export const mockMovieData = {
  results: [
    {
      original_language: 'en',
      original_title: 'Jumanji: Welcome to the Jungle',
      overview:
        "The tables are turned as four teenagers are sucked into Jumanji's world - pitted against rhinos, black mambas and an endless variety of jungle traps and puzzles. To survive, they'll play as characters from the game.",
      popularity: 101.350698,
      poster_path: '/bXrZ5iHBEjH7WMidbUDQ0U2xbmr.jpg',
      release_date: '2017-12-09',
      title: 'Jumanji: Welcome to the Jungle',
      video: false,
      vote_average: 6.5,
      vote_count: 3115
    }
  ]
};

export const cleanedMovie = [
  {
    date: 'Release Date: 2017-12-09',
    id: 101.350698,
    image: '/bXrZ5iHBEjH7WMidbUDQ0U2xbmr.jpg',
    overview:
      "Summary: The tables are turned as four teenagers are sucked into Jumanji's world - pitted against rhinos, black mambas and an endless variety of jungle traps and puzzles. To survive, they'll play as characters from the game.",
    title: 'Jumanji: Welcome to the Jungle',
    vote: 'Rating: 6.5'
  }
];

export const movieInfo = {
  movie_id: '1',
  user_id: '24',
  title: 'hello',
  poster_path: '/hello',
  release_date: '04-01-18',
  vote_average: '-5',
  overview: 'Its not real'
};

export const mockUser = {
  id: '5',
  name: 'r',
  email: 'r@r.r',
  password: 'r'
};

export const mockFavorite = {
  date: 'Release Date: 2017-12-09',
  movie_id: 101.350698,
  image: '/bXrZ5iHBEjH7WMidbUDQ0U2xbmr.jpg',
  overview:
    "Summary: The tables are turned as four teenagers are sucked into Jumanji's world - pitted against rhinos, black mambas and an endless variety of jungle traps and puzzles. To survive, they'll play as characters from the game.",
  title: 'Jumanji: Welcome to the Jungle',
  vote: 'Rating: 6.5'
};

export const mockRemainingFavorite = {
  date: 'Release Date: 2018-10-5',
  movie_id: 10,
  image: '/bXrZ5iHBEjH7WMidbUDQ0U2xbmr.jpg',
  overview: 'Summary: The tables are turned as four teenagers.',
  title: 'Fantastic Fours',
  vote: 'Rating: 2.5'
};

export const mockFavoritesArray = [
  {
    date: 'Release Date: 2018-10-5',
    movie_id: 10,
    image: '/bXrZ5iHBEjH7WMidbUDQ0U2xbmr.jpg',
    overview: 'Summary: The tables are turned as four teenagers.',
    title: 'Fantastic Fours',
    vote: 'Rating: 2.5'
  },
  {
    date: 'Release Date: 2017-12-09',
    movie_id: 101.350698,
    image: '/bXrZ5iHBEjH7WMidbUDQ0U2xbmr.jpg',
    overview:
      "Summary: The tables are turned as four teenagers are sucked into Jumanji's world - pitted against rhinos, black mambas and an endless variety of jungle traps and puzzles. To survive, they'll play as characters from the game.",
    title: 'Jumanji: Welcome to the Jungle',
    vote: 'Rating: 6.5'
  }
];
