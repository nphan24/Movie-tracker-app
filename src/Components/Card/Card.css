.card {
  display: inline-block;
  width: 250px;
  min-width: 200px;
  height: 350px;
  color: white;
  margin: 10px;
  vertical-align: top;
  background-size: 100px 100px;
  border: 2px solid black;
  transition: all .2s ease-in-out;
}

.card:hover {
  position: relative;
  z-index: 100;
  transform: scale(1.05);
}

.show-more-summary {
  width: 80px;
  height: 30px;
  font-size: 1em;
  border-radius: 30px 30px 30px 30px;
  margin: 20px 100px 0 5px;
  background-color: #E8CC77;
  color: #0A2D73;
  font-weight: 400;
  border: 2px solid white;
  text-decoration: none;
  padding: 3px 3px;
  vertical-align: center;
}

.favorites-button {
  width: 50px;
  height: 50px;
  font-size: 1em;
  border-radius: 30px 30px 30px 30px;
  background-color: white;
  color: #0A2D73;
  border: 2px solid #0A2D73;
  font-size: 1.5em;
}

.favorites-button:hover {
  background-color: #0A2D73;
  color: #E8CC77;
}

.selected .favorites-button{
  background-color: #E8CC77;
}

.show-more-summary:hover {
  background-color: #0A2D73;
  color: #E8CC77;
}


