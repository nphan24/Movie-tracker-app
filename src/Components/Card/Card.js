import React from 'react';
import { connect } from 'react-redux';
import { postToFavorites } from '../../ApiCalls/postToFavorites';
import * as Actions from '../../Actions/';
import { Link } from 'react-router-dom';
import { deleteFromFavorites } from '../../ApiCalls/deleteFromFavorites';
import './Card.css';
import PropTypes from 'prop-types';

export const Card = ({
  movie,
  user,
  favorites,
  promptSignin,
  addFavorite,
  removeFavorite,
  movieType,
  cardClass
}) => {
  const { 
    title, 
    overview, 
    release_date, 
    vote_average, 
    poster_path 
  } = movie;

  const toggleFavorite = async movie => {
    if (!user.name) {
      const error = 'Please Sign In to Add Favorites';
      promptSignin(error);
    } else if (!favorites.find(fav => fav.movie_id === movie.movie_id)) {
      await postToFavorites({ ...movie, user_id: user.id });
      addFavorite(movie);
    } else {
      removeFavorite(movie.movie_id);
      deleteFromFavorites(movie.movie_id, user.id);
    }
  };

  return (
    <article
      className={`card ${cardClass}`}
      style={{
        background: `url('https://image.tmdb.org/t/p/w500/${poster_path}')`
      }}
    >
      <h1>{title}</h1>
      <Link to={`/${movieType}/${movie.movie_id}`}>Show More</Link>
      <button onClick={() => toggleFavorite(movie)}>Favorite</button>
    </article>
  );
};

export const mapStateToProps = state => ({
  user: state.user,
  favorites: state.favorites
});

export const mapDispatchToProps = dispatch => ({
  addFavorite: movie => dispatch(Actions.addFavorite(movie)),
  removeFavorite: id => dispatch(Actions.removeFavorite(id)),
  promptSignin: error => dispatch(Actions.promptSignIn(error))
});

Card.propTypes = {
  movie: PropTypes.object,
  user: PropTypes.object,
  favorites: PropTypes.array,
  promptSignin: PropTypes.func,
  addFavorite: PropTypes.func,
  removeFavorite: PropTypes.func
};

export default connect(mapStateToProps, mapDispatchToProps)(Card);
