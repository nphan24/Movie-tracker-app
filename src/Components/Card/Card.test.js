import { Card, mapDispatchToProps, mapStateToProps } from './Card';
import * as mock from '../../ApiCalls/unit-test/mockData';
import React from 'react';
import { shallow } from 'enzyme';
import * as Actions from '../../Actions';
import { postToFavorites } from '../../ApiCalls/postToFavorites';
import { deleteFromFavorites } from '../../ApiCalls/deleteFromFavorites';
jest.mock('../../ApiCalls/postToFavorites');
jest.mock('../../ApiCalls/deleteFromFavorites');

describe('Card', () => {
  let wrapper;
  let mockMovie;
  let toggleFavorite;
  let user;
  let favorites;
  let addFavorite;
  let promptSignin;
  let removeFavorite;
  let mockFav;
  beforeEach(() => {
    promptSignin = jest.fn();
    removeFavorite = jest.fn();
    user = mock.mockUser;
    addFavorite = jest.fn();
    mockFav = mock.mockRemainingFavorite;
    favorites = mock.mockFavoritesArray;
    toggleFavorite = jest.fn();
    mockMovie = mock.movieInfo;
    wrapper = shallow(
      <Card
        addFavorite={addFavorite}
        favorites={favorites}
        movie={mockMovie}
        user={user}
      />
    );
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('toggleFavorite', () => {
    beforeEach(() => {
      wrapper.find('button').simulate('click');
    });
    it('should call postToFavorites', () => {
      const expected = { ...mockMovie, user_id: user.id };
      expect(postToFavorites).toHaveBeenCalledWith(expected);
    });

    it('should call addFavorite', () => {
      const expected = mockMovie;
      expect(addFavorite).toHaveBeenCalledWith(expected);
    });

    it('should call promptSignin if no user is present', () => {
      wrapper = shallow(
        <Card
          addFavorite={addFavorite}
          favorites={favorites}
          movie={mockMovie}
          promptSignin={promptSignin}
          user={{}}
        />
      );
      const expected = 'Please Sign In to Add Favorites';
      wrapper.find('button').simulate('click');

      expect(promptSignin).toHaveBeenCalledWith(expected);
    });

    it('should call removeFavorite', () => {
      wrapper = shallow(
        <Card
          addFavorite={addFavorite}
          favorites={favorites}
          movie={mockFav}
          removeFavorite={removeFavorite}
          promptSignin={promptSignin}
          user={user}
        />
      );
      const expected = mockFav.movie_id;
      wrapper.find('button').simulate('click');
      expect(removeFavorite).toHaveBeenCalledWith(expected);
    });

    it('should call deleteFromFavorites', () => {
      wrapper = shallow(
        <Card
          addFavorite={addFavorite}
          favorites={favorites}
          movie={mockFav}
          removeFavorite={removeFavorite}
          promptSignin={promptSignin}
          user={user}
        />
      );
      const expected = mockFav.movie_id;
      wrapper.find('button').simulate('click');
      expect(deleteFromFavorites).toHaveBeenCalledWith(expected, user.id);
    });
  });
});

describe('mapStateToProps', () => {
  it('correctly maps users to props ', () => {
    const user = mock.mockUser;
    const expected = user;
    const mockState = { user };
    const mapped = mapStateToProps(mockState);
    expect(mapped.user).toEqual(expected);
  });

  it('correctly maps favorites to props ', () => {
    const favorites = mock.mockFavoritesArray;
    const expected = favorites;
    const mockState = { favorites };
    const mapped = mapStateToProps(mockState);
    expect(mapped.favorites).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {
  it('should call dispatch with correct params on addFavorite', () => {
    const mockDispatch = jest.fn();
    const movie = mock.cleanedMovie;
    const expected = Actions.addFavorite(movie);
    const mapped = mapDispatchToProps(mockDispatch);
    mapped.addFavorite(movie);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });

  it('should call dispatch with correct params on removeFavorite', () => {
    const mockDispatch = jest.fn();
    const id = 2;
    const expected = Actions.removeFavorite(id);
    const mapped = mapDispatchToProps(mockDispatch);
    mapped.removeFavorite(id);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });

  it('should call dispatch with correct params on promptSignIn', () => {
    const mockDispatch = jest.fn();
    const error = 'error';
    const expected = Actions.promptSignIn(error);
    const mapped = mapDispatchToProps(mockDispatch);
    mapped.promptSignin(error);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
});
