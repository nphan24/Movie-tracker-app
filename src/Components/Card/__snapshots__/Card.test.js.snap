// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    addFavorite={[Function]}
    favorites={
        Array [
            Object {
              "date": "Release Date: 2018-10-5",
              "image": "/bXrZ5iHBEjH7WMidbUDQ0U2xbmr.jpg",
              "movie_id": 10,
              "overview": "Summary: The tables are turned as four teenagers.",
              "title": "Fantastic Fours",
              "vote": "Rating: 2.5",
            },
            Object {
              "date": "Release Date: 2017-12-09",
              "image": "/bXrZ5iHBEjH7WMidbUDQ0U2xbmr.jpg",
              "movie_id": 101.350698,
              "overview": "Summary: The tables are turned as four teenagers are sucked into Jumanji's world - pitted against rhinos, black mambas and an endless variety of jungle traps and puzzles. To survive, they'll play as characters from the game.",
              "title": "Jumanji: Welcome to the Jungle",
              "vote": "Rating: 6.5",
            },
          ]
    }
    movie={
        Object {
            "movie_id": "1",
            "overview": "Its not real",
            "poster_path": "/hello",
            "release_date": "04-01-18",
            "title": "hello",
            "user_id": "24",
            "vote_average": "-5",
          }
    }
    user={
        Object {
            "email": "r@r.r",
            "id": "5",
            "name": "r",
            "password": "r",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Link
          className="show-more-summary"
          replace={false}
          to="/undefined/1"
>
          Show More
</Link>,
        <button
          className="favorites-button"
          onClick={[Function]}
>
          <span
                    aria-labelledby="favorites-button"
                    role="img"
          >
                    ⭐
          </span>
</button>,
      ],
      "className": "card undefined",
      "style": Object {
        "background": "url('https://image.tmdb.org/t/p/w500//hello')",
        "backgroundSize": "250px 350px",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Show More",
          "className": "show-more-summary",
          "replace": false,
          "to": "/undefined/1",
        },
        "ref": null,
        "rendered": "Show More",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <span
            aria-labelledby="favorites-button"
            role="img"
>
            ⭐
</span>,
          "className": "favorites-button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "aria-labelledby": "favorites-button",
            "children": "⭐",
            "role": "img",
          },
          "ref": null,
          "rendered": "⭐",
          "type": "span",
        },
        "type": "button",
      },
    ],
    "type": "article",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Link
            className="show-more-summary"
            replace={false}
            to="/undefined/1"
>
            Show More
</Link>,
          <button
            className="favorites-button"
            onClick={[Function]}
>
            <span
                        aria-labelledby="favorites-button"
                        role="img"
            >
                        ⭐
            </span>
</button>,
        ],
        "className": "card undefined",
        "style": Object {
          "background": "url('https://image.tmdb.org/t/p/w500//hello')",
          "backgroundSize": "250px 350px",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Show More",
            "className": "show-more-summary",
            "replace": false,
            "to": "/undefined/1",
          },
          "ref": null,
          "rendered": "Show More",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span
              aria-labelledby="favorites-button"
              role="img"
>
              ⭐
</span>,
            "className": "favorites-button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "aria-labelledby": "favorites-button",
              "children": "⭐",
              "role": "img",
            },
            "ref": null,
            "rendered": "⭐",
            "type": "span",
          },
          "type": "button",
        },
      ],
      "type": "article",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
